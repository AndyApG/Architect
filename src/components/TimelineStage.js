import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import anime from 'animejs';
import { Icon, animation } from 'network-canvas-ui';

class TimelineStage extends PureComponent {
  static propTypes = {
    type: PropTypes.string.isRequired,
    onEditStage: PropTypes.func.isRequired,
    onEditSkipLogic: PropTypes.func.isRequired,
  };

  onMouseEnterSkipLogic = () => {
    const el = this.track;
    const rect = el.getBoundingClientRect();
    const offset = rect.width + rect.width + rect.height;

    el.setAttribute('stroke-dasharray', offset);
    el.setAttribute('stroke-dashoffset', offset);

    this.animateSkip = anime({
      targets: el,
      opacity: [0, 1],
      strokeDashoffset: [offset, 0],
      duration: animation.duration.standard,
      easing: 'easeInOutSine',
      autoplay: false,
    });

    this.animateSkip.restart();
  }

  onMouseLeaveSkipLogic = () => {
    if (!this.animateSkip) { return; }

    this.animateSkip.reverse();
    this.animateSkip.play();
  }

  snapshotSrc() {
    return `/images/timeline/stage--${this.props.type}.png`;
  }

  render() {
    const {
      onEditStage,
      onEditSkipLogic,
    } = this.props;

    return (
      <div className="timeline-stage">
        <div
          className="timeline-stage__preview"
          role="button"
          onClick={onEditStage}
          tabIndex="0"
        >
          <img src={this.snapshotSrc()} alt="" className="timeline-stage__preview-image" />
        </div>

        <svg
          preserveAspectRatio="none"
          className="timeline-stage__skip-logic-track"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 100 100"
        >
          <path
            ref={(track) => { this.track = track; }}
            d="M 0 0 L 100 0 L 100 100 L 0 100"
            vectorEffect="non-scaling-stroke"
          />
        </svg>

        <div
          className="timeline-stage__skip-logic-button"
          onMouseEnter={this.onMouseEnterSkipLogic}
          onMouseLeave={this.onMouseLeaveSkipLogic}
          onClick={onEditSkipLogic}
        >
          <Icon name="error" className="timeline-stage__skip-logic-button-icon" />
        </div>
      </div>
    );
  }
}

export default TimelineStage;
