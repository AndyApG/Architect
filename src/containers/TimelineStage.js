import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Group, Image, Rect, Text } from 'react-konva';

class TimelineStage extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      image: null,
    };
  }

  componentDidMount() {
    const image = new window.Image();
    const imageSrc = `/images/timeline/stage--${this.props.type}.png`;
    image.src = imageSrc;
    image.onload = () => {
      this.setState({
        image,
      });
    };
  }

  render() {
    const image = this.state.image;

    const {
      width,
      height,
    } = this.props;

    const centerX = width / 2;
    const snapshotHeight = height * (3 / 4);
    const snapshotWidth = snapshotHeight * (4 / 3);
    const snapshotX = centerX - (snapshotWidth / 2);

    const groupProps = {
      y: 0,
      x: 0,
      width,
      height,
    };

    const snapshotProps = {
      y: 0,
      x: snapshotX,
      height: snapshotHeight,
      width: snapshotWidth,
    };

    const addProps = {
      y: snapshotHeight + 20,
      x: centerX - 20,
      fill: 'white',
      height: 40,
      width: 40,
    };

    const ruleProps = {
      y: height / 2,
      x: snapshotX + snapshotWidth + 20,
      fill: 'white',
      height: 40,
      width: 40,
    };

    return (
      <Group
        {...groupProps}
      >
        <Text
          fill="white"
          x={0}
          y={0}
          height={20}
          width={200}
          text={this.props.title}
          fontSize={30}
          fontFamily="Calibri"
        />

        <Image
          image={image}
          {...snapshotProps}
          onClick={this.props.editStage}
        />

        <Rect {...addProps} onClick={this.props.addStage} />
        <Rect {...ruleProps} onClick={this.props.editSkip} />
      </Group>
    );
  }
}

TimelineStage.propTypes = {
  type: PropTypes.string.isRequired,
  width: PropTypes.number,
  height: PropTypes.number,
  title: PropTypes.title,
  addStage: PropTypes.func,
  editStage: PropTypes.func,
  editSkip: PropTypes.func,
};

TimelineStage.defaultProps = {
  width: 0,
  height: 0,
  title: '',
  addStage: () => {},
  editStage: () => {},
  editSkip: () => {},
};

export default TimelineStage;
