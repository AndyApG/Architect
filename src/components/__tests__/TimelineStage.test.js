/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import TimelineStage from '../TimelineStage';

const mockProps = {
  type: 'Foo',
  onEditStage: () => {},
  onEditSkipLogic: () => {},
};

describe('<TimelineStage />', () => {
  it('can render', () => {
    const component = shallow(<TimelineStage {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
