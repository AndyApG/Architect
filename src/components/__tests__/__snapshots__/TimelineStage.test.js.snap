// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimelineStage /> can render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimelineStage
    onEditSkipLogic={[Function]}
    onEditStage={[Function]}
    type="Foo"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="timeline-stage__preview"
          onClick={[Function]}
          role="button"
          tabIndex="0"
>
          <img
                    alt=""
                    className="timeline-stage__preview-image"
                    src="/images/timeline/stage--Foo.png"
          />
</div>,
        <TimelineEditSkipLogic
          onEditSkipLogic={[Function]}
/>,
      ],
      "className": "timeline-stage",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": <img
            alt=""
            className="timeline-stage__preview-image"
            src="/images/timeline/stage--Foo.png"
/>,
          "className": "timeline-stage__preview",
          "onClick": [Function],
          "role": "button",
          "tabIndex": "0",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "alt": "",
            "className": "timeline-stage__preview-image",
            "src": "/images/timeline/stage--Foo.png",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "onEditSkipLogic": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="timeline-stage__preview"
            onClick={[Function]}
            role="button"
            tabIndex="0"
>
            <img
                        alt=""
                        className="timeline-stage__preview-image"
                        src="/images/timeline/stage--Foo.png"
            />
</div>,
          <TimelineEditSkipLogic
            onEditSkipLogic={[Function]}
/>,
        ],
        "className": "timeline-stage",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": <img
              alt=""
              className="timeline-stage__preview-image"
              src="/images/timeline/stage--Foo.png"
/>,
            "className": "timeline-stage__preview",
            "onClick": [Function],
            "role": "button",
            "tabIndex": "0",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "alt": "",
              "className": "timeline-stage__preview-image",
              "src": "/images/timeline/stage--Foo.png",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "onEditSkipLogic": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
